File: samples/client/3_1_0_unit_test/java/src/test/java/org/openapijsonschematools/client/components/schemas/MultipleDependentsRequiredTest.java
package org.openapijsonschematools.client.components.schemas;

import org.junit.Assert;
import org.junit.Test;
import org.openapijsonschematools.client.configurations.JsonSchemaKeywordFlags;
import org.openapijsonschematools.client.configurations.SchemaConfiguration;
import org.openapijsonschematools.client.exceptions.ValidationException;
import org.openapijsonschematools.client.schemas.validation.MapUtils;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.AbstractMap;

public class MultipleDependentsRequiredTest {
    static final SchemaConfiguration configuration = new SchemaConfiguration(JsonSchemaKeywordFlags.onlyFormat());

    @Test
    public void testNondependantsPasses() throws ValidationException {
        // nondependants
        final var schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance();
        schema.validate(
            MapUtils.makeMap(
                new AbstractMap.SimpleEntry<String, Number>(
                    "foo",
                    1
                ),
                new AbstractMap.SimpleEntry<String, Number>(
                    "bar",
                    2
                )
            ),
            configuration
        );
    }

    @Test
    public void testMissingOtherDependencyFails() {
        // missing other dependency
        final var schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance();
        try {
            schema.validate(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Number>(
                        "bar",
                        1
                    ),
                    new AbstractMap.SimpleEntry<String, Number>(
                        "quux",
                        2
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testWithDependenciesPasses() throws ValidationException {
        // with dependencies
        final var schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance();
        schema.validate(
            MapUtils.makeMap(
                new AbstractMap.SimpleEntry<String, Number>(
                    "foo",
                    1
                ),
                new AbstractMap.SimpleEntry<String, Number>(
                    "bar",
                    2
                ),
                new AbstractMap.SimpleEntry<String, Number>(
                    "quux",
                    3
                )
            ),
            configuration
        );
    }

    @Test
    public void testMissingBothDependenciesFails() {
        // missing both dependencies
        final var schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance();
        try {
            schema.validate(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Number>(
                        "quux",
                        1
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testMissingDependencyFails() {
        // missing dependency
        final var schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance();
        try {
            schema.validate(
                MapUtils.makeMap(
                    new AbstractMap.SimpleEntry<String, Number>(
                        "foo",
                        1
                    ),
                    new AbstractMap.SimpleEntry<String, Number>(
                        "quux",
                        2
                    )
                ),
                configuration
            );
            throw new RuntimeException("A different exception must be thrown");
        } catch (ValidationException ignored) {
            ;
        }
    }

    @Test
    public void testNeitherPasses() throws ValidationException {
        // neither
        final var schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance();
        schema.validate(
            MapUtils.makeMap(
            ),
            configuration
        );
    }
}

File: samples/client/3_1_0_unit_test/kotlin/src/test/kotlin/org/openapijsonschematools/client/components/schemas/MultipleDependentsRequiredTest.kt
package org.openapijsonschematools.client.components.schemas

import org.openapijsonschematools.client.configurations.JsonSchemaKeywordFlags
import org.openapijsonschematools.client.configurations.SchemaConfiguration
import org.openapijsonschematools.client.exceptions.ValidationException

import kotlin.test.Test
import kotlin.test.assertFailsWith

class MultipleDependentsRequiredTest {
    companion object {
        val configuration = SchemaConfiguration(JsonSchemaKeywordFlags.Builder().format().build())
    }

    @Test
    fun testNondependantsPasses() {
        // nondependants
        val schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance()
        schema.validate(
            mapOf(
                Pair(
                    "foo",
                    1
                ),
                Pair(
                    "bar",
                    2
                )
            ),
            configuration
        )
    }

    @Test
    fun testMissingOtherDependencyFails() {
        // missing other dependency
        val schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "bar",
                            1
                        ),
                        Pair(
                            "quux",
                            2
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testWithDependenciesPasses() {
        // with dependencies
        val schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance()
        schema.validate(
            mapOf(
                Pair(
                    "foo",
                    1
                ),
                Pair(
                    "bar",
                    2
                ),
                Pair(
                    "quux",
                    3
                )
            ),
            configuration
        )
    }

    @Test
    fun testMissingBothDependenciesFails() {
        // missing both dependencies
        val schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "quux",
                            1
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testMissingDependencyFails() {
        // missing dependency
        val schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance()
        assertFailsWith<ValidationException>(
            block = {
                schema.validate(
                    mapOf(
                        Pair(
                            "foo",
                            1
                        ),
                        Pair(
                            "quux",
                            2
                        )
                    ),
                    configuration
                )
            }
        )
    }

    @Test
    fun testNeitherPasses() {
        // neither
        val schema = MultipleDependentsRequired.MultipleDependentsRequired1.getInstance()
        schema.validate(
            mapOf<Any?, Any?>(
            ),
            configuration
        )
    }
}
